polyfill/lib/ecmascript.mjs

```diff
@@ -3761,6 +3685,17 @@ export function RejectDuration(y, mon, w, d, h, min, s, ms, µs, ns) {
     const propSign = MathSign(prop);
     if (propSign !== 0 && propSign !== sign) throw new RangeError('mixed-sign values not allowed as duration fields');
   }
+  if (MathAbs(y) >= 2 ** 32 || MathAbs(mon) >= 2 ** 32 || MathAbs(w) >= 2 ** 32) {
+    throw new RangeError('years, months, and weeks must be < 2³²');
+  }
+  const msResult = TruncatingDivMod(ms, 1e3);
+  const µsResult = TruncatingDivMod(µs, 1e6);
+  const nsResult = TruncatingDivMod(ns, 1e9);
+  const remainderSec = MathTrunc((msResult.mod * 1e6 + µsResult.mod * 1e3 + nsResult.mod) / 1e9);
+  const totalSec = d * 86400 + h * 3600 + min * 60 + s + msResult.div + µsResult.div + nsResult.div + remainderSec;
```

Comment:
I was optimistic that this could still be calculated correctly with floating point arithmetic, using `std::remquo`. I've got something that works for milliseconds, which I'll push to this branch in case anyone else can spot issues with it. But even the 3 extra bits of accuracy that you get from `std::remquo` seems to be just not enough for dividing microseconds by 10⁶ or nanoseconds by 10⁹.

I suspect that implementations could theoretically use the numerical technique from `std::remquo` to get more bits of accuracy, but that seems like an unreasonable burden. So maybe we should change course. I have three ideas.

1. Specify that this particular calculation should be performed in the 𝔽 domain, exactly as if you were to do `Number.isSafeInteger(d * 86400 + h * 3600 + min * 60 + s + Math.trunc(ms / 1e3 + µs / 1e6 + ns / 1e9))` in JS. This would produce both false positives and false negatives as you have shown, but only when `ms`, `µs`, or `ns` are already unsafe integers and the total is just under or over the limit.
2. Add an additional limit on user code, that besides the combined limit, each component individually must be a safe integer. This is a departure from how we've previously thought about the Duration API, where using unsafe integers means you accept a certain loss in precision when converting between units.
3. Restrict the total length of the duration to 1.728×10¹³ s (the difference between the earliest and latest possible `Temporal.Instant`s), instead of `MAX_SAFE_INTEGER` as we have currently decided. This should allow the division to be implemented naively using floating point arithmetic exactly as it was written in the spec, because assuming an error of 0.5 ULP each division will be off by at most 9765 ns, which will not affect the truncate operation.

Any opinions on this? I'm open to being convinced of any of them.

Reply:
> 3\. Restrict the total length of the duration to 1.728×10¹³ s (the difference between the earliest and latest possible `Temporal.Instant`s), instead of `MAX_SAFE_INTEGER` as we have currently decided.

Do you mean the time portion of the duration or the entire duration including years?  I'm asking because, if the latter, then ~550K years means we won't need 32-bit years or months. 

> 2\. Add an additional limit on user code, that besides the combined limit, each component individually must be a safe integer.

Do you mean that currently you can use  `{ nanoseconds:2^60 }`  but with (2) you couldn't?

Reply:
In the meantime I did actually find freely licensed implementations of `std::fma()` and `std::remquo()` (with sufficient quotient precision), in [OpenLibm](https://openlibm.org/). I made a line-by-line port of them to JS, which is [here](https://gist.github.com/ptomato/d1ac949bbd3e5719766e770b7c0415bb).

However I also realized that because we are only multiplying and dividing by powers of 10, you can actually perform these `fma` and `remquo` operations without loss of precision, simply by string manipulation (knocking digits off the end of the string, or sticking them back on). This is what I've now pushed in this PR, to run the test262 tests against.

For a C++ implementation, I'd recommend using the standard library's `std::fma()`, and copying the extra-precise `remquo()` implementation from OpenLibm. So I take back my earlier opinion that this is an unreasonable burden on implementations. I would no longer suggest changing anything, but if we did want to simplify then I'd suggest (1).

Re. (2),
> Do you mean that currently you can use { nanoseconds:2^60 } but with (2) you couldn't?

Yes, that's right.

Re. (3),
> Do you mean the time portion of the duration or the entire duration including years?

Just the time portion.

Reply:
> > Do you mean that currently you can use { nanoseconds:2^60 } but with (2) you couldn't?
> 
> Yes, that's right.

FWIW, it seems OK to apply this limit. There is no use case I'm aware of for measuring durations > 500K years in any other unit beside years.